hibernate-core/src/main/java/org/hibernate/id/enhanced/PooledLoThreadLocalOptimizer.java

```diff
@@ -58,26 +58,28 @@ public PooledLoThreadLocalOptimizer(Class returnClass, int incrementSize) {
 	public Serializable generate(AccessCallback callback) {
 		if ( callback.getTenantIdentifier() == null ) {
 			final GenerationState local = localAssignedIds.get();
-			if ( local.value != null && local.value.lt( local.upperLimitValue ) ) {
-				return local.value.makeValueThenIncrement();
-			}
```

Comment:
Can you help me understand this commit? The comment suggests you want to avoid a `synchronization`, but this block was already unsynched?

Reply:
If you look closely, when the condition was `false`, we would pass through this block and reach the synchronized` block, even for single-tenancy. This is what I'm changing.

Basically we used to synchronize whenever we needed to call the database to renew the pool. Which is completely unnecessary. More importantly, it *may* be a source of deadlock in some situations. Simplified example:

1. Thread 1 (T1) already has a connection (from a pool of size 1)
2. Thread 2 (T2) doesn't have a connection.
3. Thread 2 requests a new ID, enters the synchronization block and waits for a connection from T1.
4. Thread 1 requests a new ID, tries to enter the synchronization block and waits for T2 to leave it.
5. T1 waits for T2, T2 waits for T1: deadlock... until T2 times out waiting for a connection.

Admittedly that's an extreme scenario. Still, I think it's a valid concern that could happen even with a pool of a normal size.

To be more explicit, here is the code before:

```
if ( single-tenant ) {
  if ( we have more values in the pool ) {
    return the next value from the pool
  }
  else {
    synchronized (this) {
      get the next low value from the sequence
      return the next value from the pool
    }
  }
}
else { // multi-tenant
  synchronized (this) {
    get the pool for this tenant
    if ( we have more values in the pool ) {
      return the next value from the pool
    }
    else {
        get the next low value from the sequence
        return the next value from the pool
    }
  }
}
```

And here is the code after this change

```
if ( single-tenant ) {
  if ( we have more values in the pool ) {
    return the next value from the pool
  }
  else { // NO SYNCHRONIZATION HERE
    get the next low value from the sequence
    return the next value from the pool
  }
}
else { // multi-tenant
  synchronized (this) {
    get the pool for this tenant
    if ( we have more values in the pool ) {
      return the next value from the pool
    }
    else {
        get the next low value from the sequence
        return the next value from the pool
    }
  }
}
```

Reply:
ah, many thanks - I missed the case when reading yesterday. It's great you spotted this!